compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3358,6 +3358,8 @@ object Types {
           absMems.find(_.symbol.name == nme.apply)
         else None
       }
+      else if (tp.isInstanceOf[HKApply])
+        unapply(tp.asInstanceOf[HKApply].superType)
       else None
```

Comment:
Good diagnosis to fix the bug! For the fix, can you try instead to add a case to `zeroParamClass` above?

    case tp: HKApply => zeroParamClass(tp.superType)

That would be more regular.